src/main/java/filter/Parser.java

```diff
@@ -17,24 +17,62 @@
 
 public final class Parser {
 
-    private Parser(ArrayList<Token> input) {
+    private final List<Token> input;
+
+    /**
+     * Determines if the parser should fail gracefully if it finds the input
+     * in a incomplete state. See {@link #check}.
+     */
+    private final boolean isInCheckMode;
```

Comment:
Better name and/or comment? We should explain somewhere what the 'check mode' means.
The comments leans a bit towards 'how it is used' rather than 'what it means'.